File: library/src/main/java/com/henorek/discharge/solutions/builders/DischargeDialogBuilder.java
package com.henorek.discharge.solutions.builders;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.database.Cursor;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListAdapter;

public class DischargeDialogBuilder {

  private AlertDialog.Builder dialog;

  DischargeDialogBuilder(final Context context) {
    dialog = new AlertDialog.Builder(context);
  }

  public DischargeDialogBuilder withTitle(int titleResourceId) {
    dialog.setTitle(titleResourceId);
    return this;
  }

  public DischargeDialogBuilder withTitle(CharSequence title) {
    dialog.setTitle(title);
    return this;
  }

  public DischargeDialogBuilder withCustomTitle(View customTitleView) {
    dialog.setCustomTitle(customTitleView);
    return this;
  }

  public DischargeDialogBuilder withMessage(int messageResourceId) {
    dialog.setMessage(messageResourceId);
    return this;
  }

  public DischargeDialogBuilder withMessage(CharSequence message) {
    dialog.setMessage(message);
    return this;
  }

  public DischargeDialogBuilder withIcon(int iconResourceId) {
    dialog.setIcon(iconResourceId);
    return this;
  }

  public DischargeDialogBuilder withIcon(Drawable icon) {
    dialog.setIcon(icon);
    return this;
  }

  public DischargeDialogBuilder withIconAttribute(int attrResourceId) {
    if (Build.VERSION.SDK_INT > 11) dialog.setIconAttribute(attrResourceId);
    return this;
  }

  public DischargeDialogBuilder withPositiveButton(int textResourceId, DialogInterface.OnClickListener listener) {
    dialog.setPositiveButton(textResourceId, listener);
    return this;
  }

  public DischargeDialogBuilder withPositiveButton(CharSequence text, DialogInterface.OnClickListener listener) {
    dialog.setPositiveButton(text, listener);
    return this;
  }

  public DischargeDialogBuilder withNegativeButton(int textResourceId, DialogInterface.OnClickListener listener) {
    dialog.setNegativeButton(textResourceId, listener);
    return this;
  }

  public DischargeDialogBuilder withNegativeButton(CharSequence text, DialogInterface.OnClickListener listener) {
    dialog.setNegativeButton(text, listener);
    return this;
  }

  public DischargeDialogBuilder withNeutralButton(int textResourceId, DialogInterface.OnClickListener listener) {
    dialog.setNeutralButton(textResourceId, listener);
    return this;
  }

  public DischargeDialogBuilder withNeutralButton(CharSequence text, DialogInterface.OnClickListener listener) {
    dialog.setNeutralButton(text, listener);
    return this;
  }

  public DischargeDialogBuilder withCancelable(boolean cancelable) {
    dialog.setCancelable(cancelable);
    return this;
  }

  public DischargeDialogBuilder withOnCancelListener(DialogInterface.OnCancelListener onCancelListener) {
    dialog.setOnCancelListener(onCancelListener);
    return this;
  }

  public DischargeDialogBuilder withOnDismissListener(DialogInterface.OnDismissListener onDismissListener) {
    if (Build.VERSION.SDK_INT > 17) dialog.setOnDismissListener(onDismissListener);
    return this;
  }

  public DischargeDialogBuilder withOnKeyListener(DialogInterface.OnKeyListener onKeyListener) {
    dialog.setOnKeyListener(onKeyListener);
    return this;
  }

  public DischargeDialogBuilder withItems(int itemsResourceId, DialogInterface.OnClickListener listener) {
    dialog.setItems(itemsResourceId, listener);
    return this;
  }

  public DischargeDialogBuilder withItems(CharSequence[] items, DialogInterface.OnClickListener listener) {
    dialog.setItems(items, listener);
    return this;
  }

  public DischargeDialogBuilder withAdapter(ListAdapter adapter, DialogInterface.OnClickListener listener) {
    dialog.setAdapter(adapter, listener);
    return this;
  }

  public DischargeDialogBuilder withCursor(Cursor cursor, DialogInterface.OnClickListener listener,
      String labelColumn) {
    dialog.setCursor(cursor, listener, labelColumn);
    return this;
  }

  public DischargeDialogBuilder withMultiChoiceItems(int itemsId, boolean[] checkedItems,
      DialogInterface.OnMultiChoiceClickListener listener) {
    dialog.setMultiChoiceItems(itemsId, checkedItems, listener);
    return this;
  }

  public DischargeDialogBuilder withMultiChoiceItems(CharSequence[] items, boolean[] checkedItems,
      DialogInterface.OnMultiChoiceClickListener listener) {
    dialog.setMultiChoiceItems(items, checkedItems, listener);
    return this;
  }

  public DischargeDialogBuilder withMultiChoiceItems(Cursor cursor, String isCheckedColumn, String labelColumn,
      DialogInterface.OnMultiChoiceClickListener listener) {
    dialog.setMultiChoiceItems(cursor, isCheckedColumn, labelColumn, listener);
    return this;
  }

  public DischargeDialogBuilder withSingleChoiceItems(int itemsId, int checkedItem,
      DialogInterface.OnClickListener listener) {
    dialog.setSingleChoiceItems(itemsId, checkedItem, listener);
    return this;
  }

  public DischargeDialogBuilder withSingleChoiceItems(Cursor cursor, int checkedItem, String labelColumn,
      DialogInterface.OnClickListener listener) {
    dialog.setSingleChoiceItems(cursor, checkedItem, labelColumn, listener);
    return this;
  }

  public DischargeDialogBuilder withSingleChoiceItems(CharSequence[] items, int checkedItem,
      DialogInterface.OnClickListener listener) {
    dialog.setSingleChoiceItems(items, checkedItem, listener);
    return this;
  }

  public DischargeDialogBuilder withSingleChoiceItems(ListAdapter adapter, int checkedItem,
      DialogInterface.OnClickListener listener) {
    dialog.setSingleChoiceItems(adapter, checkedItem, listener);
    return this;
  }

  public DischargeDialogBuilder withOnItemSelectedListener(AdapterView.OnItemSelectedListener listener) {
    dialog.setOnItemSelectedListener(listener);
    return this;
  }

  public DischargeDialogBuilder withView(int layoutResId) {
    if (Build.VERSION.SDK_INT > 21) dialog.setView(layoutResId);
    return this;
  }

  public DischargeDialogBuilder withView(View view) {
    dialog.setView(view);
    return this;
  }

  public DischargeDialog build() {
    return new DischargeDialog(dialog.create());
  }
}

File: library/src/main/java/com/henorek/discharge/solutions/builders/DischargeDialogBuilder.kt
package com.henorek.discharge.solutions.builders

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.database.Cursor
import android.graphics.drawable.Drawable
import android.os.Build
import android.view.View
import android.widget.AdapterView
import android.widget.ListAdapter

class DischargeDialogBuilder internal constructor(context: Context) {

  private val dialog: AlertDialog.Builder

  init {
    dialog = AlertDialog.Builder(context)
  }

  fun withTitle(titleResourceId: Int): DischargeDialogBuilder {
    dialog.setTitle(titleResourceId)
    return this
  }

  fun withTitle(title: CharSequence): DischargeDialogBuilder {
    dialog.setTitle(title)
    return this
  }

  fun withCustomTitle(customTitleView: View): DischargeDialogBuilder {
    dialog.setCustomTitle(customTitleView)
    return this
  }

  fun withMessage(messageResourceId: Int): DischargeDialogBuilder {
    dialog.setMessage(messageResourceId)
    return this
  }

  fun withMessage(message: CharSequence): DischargeDialogBuilder {
    dialog.setMessage(message)
    return this
  }

  fun withIcon(iconResourceId: Int): DischargeDialogBuilder {
    dialog.setIcon(iconResourceId)
    return this
  }

  fun withIcon(icon: Drawable): DischargeDialogBuilder {
    dialog.setIcon(icon)
    return this
  }

  fun withIconAttribute(attrResourceId: Int): DischargeDialogBuilder {
    if (Build.VERSION.SDK_INT > 11) dialog.setIconAttribute(attrResourceId)
    return this
  }

  fun withPositiveButton(textResourceId: Int, listener: DialogInterface.OnClickListener): DischargeDialogBuilder {
    dialog.setPositiveButton(textResourceId, listener)
    return this
  }

  fun withPositiveButton(text: CharSequence, listener: DialogInterface.OnClickListener): DischargeDialogBuilder {
    dialog.setPositiveButton(text, listener)
    return this
  }

  fun withNegativeButton(textResourceId: Int, listener: DialogInterface.OnClickListener): DischargeDialogBuilder {
    dialog.setNegativeButton(textResourceId, listener)
    return this
  }

  fun withNegativeButton(text: CharSequence, listener: DialogInterface.OnClickListener): DischargeDialogBuilder {
    dialog.setNegativeButton(text, listener)
    return this
  }

  fun withNeutralButton(textResourceId: Int, listener: DialogInterface.OnClickListener): DischargeDialogBuilder {
    dialog.setNeutralButton(textResourceId, listener)
    return this
  }

  fun withNeutralButton(text: CharSequence, listener: DialogInterface.OnClickListener): DischargeDialogBuilder {
    dialog.setNeutralButton(text, listener)
    return this
  }

  fun withCancelable(cancelable: Boolean): DischargeDialogBuilder {
    dialog.setCancelable(cancelable)
    return this
  }

  fun withOnCancelListener(onCancelListener: DialogInterface.OnCancelListener): DischargeDialogBuilder {
    dialog.setOnCancelListener(onCancelListener)
    return this
  }

  fun withOnDismissListener(onDismissListener: DialogInterface.OnDismissListener): DischargeDialogBuilder {
    if (Build.VERSION.SDK_INT > 17) dialog.setOnDismissListener(onDismissListener)
    return this
  }

  fun withOnKeyListener(onKeyListener: DialogInterface.OnKeyListener): DischargeDialogBuilder {
    dialog.setOnKeyListener(onKeyListener)
    return this
  }

  fun withItems(itemsResourceId: Int, listener: DialogInterface.OnClickListener): DischargeDialogBuilder {
    dialog.setItems(itemsResourceId, listener)
    return this
  }

  fun withItems(items: Array<CharSequence>, listener: DialogInterface.OnClickListener): DischargeDialogBuilder {
    dialog.setItems(items, listener)
    return this
  }

  fun withAdapter(adapter: ListAdapter, listener: DialogInterface.OnClickListener): DischargeDialogBuilder {
    dialog.setAdapter(adapter, listener)
    return this
  }

  fun withCursor(cursor: Cursor, listener: DialogInterface.OnClickListener,
      labelColumn: String): DischargeDialogBuilder {
    dialog.setCursor(cursor, listener, labelColumn)
    return this
  }

  fun withMultiChoiceItems(itemsId: Int, checkedItems: BooleanArray,
      listener: DialogInterface.OnMultiChoiceClickListener): DischargeDialogBuilder {
    dialog.setMultiChoiceItems(itemsId, checkedItems, listener)
    return this
  }

  fun withMultiChoiceItems(items: Array<CharSequence>, checkedItems: BooleanArray,
      listener: DialogInterface.OnMultiChoiceClickListener): DischargeDialogBuilder {
    dialog.setMultiChoiceItems(items, checkedItems, listener)
    return this
  }

  fun withMultiChoiceItems(cursor: Cursor, isCheckedColumn: String, labelColumn: String,
      listener: DialogInterface.OnMultiChoiceClickListener): DischargeDialogBuilder {
    dialog.setMultiChoiceItems(cursor, isCheckedColumn, labelColumn, listener)
    return this
  }

  fun withSingleChoiceItems(itemsId: Int, checkedItem: Int,
      listener: DialogInterface.OnClickListener): DischargeDialogBuilder {
    dialog.setSingleChoiceItems(itemsId, checkedItem, listener)
    return this
  }

  fun withSingleChoiceItems(cursor: Cursor, checkedItem: Int, labelColumn: String,
      listener: DialogInterface.OnClickListener): DischargeDialogBuilder {
    dialog.setSingleChoiceItems(cursor, checkedItem, labelColumn, listener)
    return this
  }

  fun withSingleChoiceItems(items: Array<CharSequence>, checkedItem: Int,
      listener: DialogInterface.OnClickListener): DischargeDialogBuilder {
    dialog.setSingleChoiceItems(items, checkedItem, listener)
    return this
  }

  fun withSingleChoiceItems(adapter: ListAdapter, checkedItem: Int,
      listener: DialogInterface.OnClickListener): DischargeDialogBuilder {
    dialog.setSingleChoiceItems(adapter, checkedItem, listener)
    return this
  }

  fun withOnItemSelectedListener(listener: AdapterView.OnItemSelectedListener): DischargeDialogBuilder {
    dialog.setOnItemSelectedListener(listener)
    return this
  }

  fun withView(layoutResId: Int): DischargeDialogBuilder {
    if (Build.VERSION.SDK_INT > 21) dialog.setView(layoutResId)
    return this
  }

  fun withView(view: View): DischargeDialogBuilder {
    dialog.setView(view)
    return this
  }

  fun build(): DischargeDialog {
    return DischargeDialog(dialog.create())
  }
}
